src/sync_client.rs

```diff
@@ -38,128 +40,293 @@ use crate::{
+				error!("Cannot send block verified message: {error}");
+			}
+		}
+
+		Ok(())
+	}
+	fn verify_cells(
+		&self,
+		block_num: u32,
+		dimensions: &Dimensions,
+		cells: &[Cell],
+		commitments: &[[u8; 48]],
+		public_parameters: &PublicParameters,
+	) -> Result<(Vec<Position>, Vec<Position>)>;
+}
+
+// trait SyncClone {
+//     fn clone_box(&self) -> Box<dyn SyncClient>;
+// }
+
+// impl<T> SyncClone for T
+// where
+//     T: 'static + SyncClient + Clone,
+// {
+//     fn clone_box(&self) -> Box<dyn SyncClient> {
+//         Box::new(self.clone())
+//     }
+// }
+
+#[derive(Clone)]
+pub struct SyncClientImpl {
+	// cfg: SyncClientConfig,
 	db: Arc<DB>,
-	block_number: u32,
+	// pp: PublicParameters,
+	// block_tx: Option<Sender<BlockVerified>>,
+	// block_number: u32,
 	network_client: Client,
-	pp: PublicParameters,
-	block_tx: Option<Sender<BlockVerified>>,
-) -> Result<()> {
-	if is_block_header_in_db(db.clone(), block_number)
-		.context("Failed to check if block header is in DB")?
-	{
-		// TODO: If block header storing fails, that block will be skipped upon restart
-		// Better option would be to check for confidence
-		return Ok(());
-	};
-
-	// if block header look up fails, only then comes here for
-	// fetching and storing block header as part of (light weight)
-	// syncing process
-	let begin = SystemTime::now();
-
-	let (header, header_hash) = rpc::get_header_by_block_number(rpc_client, block_number)
-		.await
-		.context("Failed to get block {block_number} by block number")?;
-
-	let HeaderExtension::V1(xt) = &header.extension;
-
-	info!(block_number, "App index {:?}", xt.app_lookup.index);
-
-	store_block_header_in_db(db.clone(), block_number, &header)
-		.context("Failed to store block header in DB")?;
-
-	info!(
-		block_number,
-		"Synced block header: \t{:?}",
-		begin.elapsed()?
-	);
-
-	// If it's found that this certain block is not verified
-	// then it'll be verified now
-	if is_confidence_in_db(db.clone(), block_number)
-		.context("Failed to check if confidence is in DB")?
-	{
-		return Ok(());
-	};
-
-	let begin = SystemTime::now();
-
-	let dimensions =
-		Dimensions::new(xt.commitment.rows, xt.commitment.cols).context("Invalid dimensions")?;
-
-	let commitments = commitments::from_slice(&xt.commitment.commitment)?;
-
-	// now this is in `u64`
-	let cell_count = rpc::cell_count_for_confidence(cfg.confidence);
-	let positions = rpc::generate_random_cells(&dimensions, cell_count);
-
-	let (dht_fetched, unfetched) = network_client
-		.fetch_cells_from_dht(block_number, &positions)
-		.await;
+	rpc_client: OnlineClient<AvailConfig>,
+}
 
-	info!(
-		block_number,
-		"Number of cells fetched from DHT: {}",
-		dht_fetched.len()
-	);
+#[async_trait]
+impl SyncClient for SyncClientImpl {
+	fn block_header_in_db(&self, block_number: u32) -> Result<bool> {
+		Ok(is_block_header_in_db(self.db.clone(), block_number)
+			.context("Failed to check if block header is in DB")?)
```

Comment:
Since we are using `context` fn, every error is wrapped in anyhow error, and we can remove `Ok({result}?)` wrapping and write only `{result}`